Handle Non-graceful Node Shutdown

sonasingh46 opened the pull request:
Signed-off-by: Ashutosh Kumar <sonasingh46@gmail.com>

Co-authored-by: Ashutosh Kumar <sonasingh46@gmail.com>


#### What type of PR is this?
Implements KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2268-non-graceful-shutdown

/kind feature

#### What this PR does / why we need it:
This PR adds a feature to detach volume immediately and does not wait for the 6 min timeout in case of a non graceful shutdown of a node that has an `node.kubernetes.io/out-of-service` taint added manually.

Adds feature gate `NodeOutOfServiceVolumeDetach` for this feature which is disabled by default. 

#### Which issue(s) this PR fixes:
Fixes #

#### Special notes for your reviewer:

Performed the following tests: 



Used kubernetes cluster created on vSphere infra and vSphere CSI driver. 
Kubernetes Version: 
```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.3", GitCommit:"816c97ab8cff8a1c72eccca1026f7820e93e0d25", GitTreeState:"clean", BuildDate:"2022-01-25T21:25:17Z", GoVersion:"go1.17.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.3", GitCommit:"816c97ab8cff8a1c72eccca1026f7820e93e0d25", GitTreeState:"clean", BuildDate:"2022-01-25T21:19:12Z", GoVersion:"go1.17.6", Compiler:"gc", Platform:"linux/amd64"}
```

Kube-controller-image was based out on the current PR

##### Test 1:

```
1. Deployed the kube-controller-manager with current code changes and disabled 
the non graceful shutdown feature gate. (It is disabled by default)

2. Created a statefulset pod. 

3. Shutdown the node on which the pod is scheduled. Node is shutdown using the 
`Shut Down Guest OS` from vSphere UI ( This shutdown is a non graceful shutdown ) 

4. Observed that after 5 mins, the pod changed to `Terminating` state. 

5. Observed that even after 6 mins, ( i.e total 6+5 = 11 mins ) the pod is stuck in `Terminating` state
```

##### Test 2:

```
1. Deployed the kube-controller-manager with current code changes and disabled 
the non graceful shutdown feature gate. (It is disabled by default).

2. Created a statefulset pod. 

3. Shutdown the node on which the pod is scheduled. Node is shutdown using the 
`Shut Down Guest OS` from vSphere UI ( This shutdown is a non graceful shutdown ) 

4. Observed that after 5 mins, the pod changed to `Terminating` state. 

5. Deleted the pod using `kubectl delete pod <pod-name> --force --grace-period 0`

6. The pod immediately got scheduled to a different healthy node but was stuck in `ContainerCreating` 
state for 6 mins. The pod came into `Running` state after 6 mins. It had the following events: 
 ----     ------                  ----   ----                     -------
  Normal   Scheduled               6m10s  default-scheduler        Successfully assigned default/test-sts-0 to k8s-node-716-1644856168
  Warning  FailedAttachVolume      6m10s  attachdetach-controller  Multi-Attach error for volume "pvc-b83ddf01-3029-4666-aea5-f43c91d8ddf0" Volume is already exclusively attached to one node and can't be attached to another
  Warning  FailedMount             4m7s   kubelet                  Unable to attach or mount volumes: unmounted volumes=[test-volume], unattached volumes=[test-volume kube-api-access-48d69]: timed out waiting for the condition
  Warning  FailedMount             113s   kubelet                  Unable to attach or mount volumes: unmounted volumes=[test-volume], unattached volumes=[kube-api-access-48d69 test-volume]: timed out waiting for the condition
  Normal   SuccessfulAttachVolume  1s     attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-b83ddf01-3029-4666-aea5-f43c91d8ddf0"
  
```

##### Test 3:

```
1. Deployed the kube-controller-manager with current code changes and enabled the 
non graceful shutdown feature gate. The feature gate can be enabled by adding the 
following to kube-controller-manager manifest yaml.
spec:
  containers:
  - command:
    // Add the below line
    - --feature-gates=NodeOutOfServiceVolumeDetach=true

2. Created a statefulset pod. 

3. Shutdown the node on which the pod is scheduled. Node is shutdown using the 
`Shut Down Guest OS` from vSphere UI ( This shutdown is a non graceful shutdown ) 

4. Observed that after 5 mins, the pod changed to `Terminating` state. 

5. Taint the node on which the pod was scheduled using the command: 
kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=value1:NoExecute

6. The pod immediately got scheduled to a different healthy and came into running state in next couple of seconds without waiting for the 6 mins detach timeout period.
  
```
#### Does this PR introduce a user-facing change?

```release-note
Non graceful node shutdown handling.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2268-non-graceful-shutdown

```

xing-yang committed:
Handle Non-graceful Node Shutdown

Signed-off-by: Ashutosh Kumar <sonasingh46@gmail.com>

Co-authored-by: Ashutosh Kumar <sonasingh46@gmail.com>

xing-yang commented:
/assign

xing-yang commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +128,77 @@ func isPodTerminated(pod *v1.Pod) bool {
+}
+
+// isPodTolerate returns true if the pod tolerates the provided taint.
+func isPodTolerate(taint *v1.Taint, pod *v1.Pod) bool {
```
Rename to "podToleratesTaint"?

xing-yang commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -221,7 +296,6 @@ func (gcc *PodGCController) checkIfNodeExists(ctx context.Context, name string)
 
 // gcUnscheduledTerminating deletes pods that are terminating and haven't been scheduled to a particular node.
 func (gcc *PodGCController) gcUnscheduledTerminating(pods []*v1.Pod) {
-	klog.V(4).Infof("GC'ing unscheduled pods which are terminating.")
```
We don't need to remove this line.

xing-yang commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -324,7 +325,7 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
 	defer adc.pvcQueue.ShutDown()
 
-	klog.Infof("Starting attach detach controller")
+	klog.Infof("XY Test Shutdown: Starting attach detach controller")
```
Remove this change.

xing-yang commented on pkg/features/kube_features.go:
```diff
@@ -826,6 +826,14 @@ const (
 	// Stop auto-generation of secret-based service account tokens.
 	LegacyServiceAccountTokenNoAutoGeneration featuregate.Feature = "LegacyServiceAccountTokenNoAutoGeneration"
+
+	// owner: @xing-yang
```
Can you also add your github id

xing-yang commented on pkg/features/kube_features.go:
```diff
@@ -826,6 +826,14 @@ const (
+	// alpha: v1.24
+	//
+	// Allow pods to failover to a different node in case of non graceful node shutdown
+	// before the 6 min timeour period
```
I think we can remove this "before the 6 min timeour period" here as it is not clear where this timeout happens with this text here.

xing-yang commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -277,6 +277,7 @@ func (rc *reconciler) mountAttachVolumes() {
 }
 
 func (rc *reconciler) unmountDetachDevices() {
+	// Test Shutdown: csi_attacher.go: UnmountDevice("/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-xxxx-xxxx-xxxxx-xxxxx/globalmount") for k8s-node-000
```
Remove this comment

xing-yang commented on pkg/util/taints/taints.go:
```diff
@@ -34,6 +34,10 @@ const (
 )
 
+const (
+	OutOfServiceTaint = "out-of-service"
```
Add a comment for this const, including a reference to the KEP.

sftim replied:
Per https://github.com/kubernetes/enhancements/issues/2268#issuecomment-1059349170, I think we should use a registered taint and _not_ a private taint here.
(In other words, the KEP is proposing an invalid design - we should fix the KEP and then update the implementation to match).

sonasingh46 replied:
@sftim -- Do we need to do any specific code changes to make a taint registered or simply following the naming conventions and putting it in the doc should work?

xing-yang replied:
We probably should add the new taint to here: https://github.com/kubernetes/kubernetes/blob/v1.24.0-alpha.3/staging/src/k8s.io/api/core/v1/well_known_taints.go#L48

xing-yang replied:
Sure, we can submit a PR to add the new taint to the dev-1.24 branch of k/website.

xing-yang commented:
/ok-to-test

xing-yang commented:
@sonasingh46 Can you add a release note?

xing-yang commented:
/assign @jingxu97 @gnufied

xing-yang commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +128,77 @@ func isPodTerminated(pod *v1.Pod) bool {
+	}
+
+	terminatedPodCount := len(terminatedPods)
+	//deleteCount := terminatedPodCount - gcc.terminatedPodThreshold
```
Remove this line.

xing-yang commented:
/assign @YuikoTakada

xing-yang commented:
/retest

YuikoTakada commented:
> #### What this PR does / why we need it:
> This PR adds a feature to detach volume immediately in case of a non graceful shutdown of a node and does not wait for the 6 min timeout.

in case of being added `out-of-service` taint from a user manually, right?

sonasingh46 commented:
@YuikoTakada 

>in case of being added out-of-service taint from a user manually, right?

Yes. Let me update the PR description. Thanks

jingxu97 commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -85,8 +88,6 @@ func NewPodGC(ctx context.Context, kubeClient clientset.Interface, podInformer c
 func (gcc *PodGCController) Run(ctx context.Context) {
 	defer utilruntime.HandleCrash()
-
-	klog.Infof("Starting GC controller")
```
why remove this log?

jingxu97 commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +128,76 @@ func isPodTerminated(pod *v1.Pod) bool {
+	if phase := pod.Status.Phase; phase != v1.PodPending && phase != v1.PodRunning && phase != v1.PodUnknown {
+		return true
+	}
+	if pod.ObjectMeta.DeletionTimestamp != nil {
```
I am thinking maybe  DeletionTimestamp is the only thing we should check here? Once node is not ready, we depends on eviction to mark eletiontimestamps on every pod?

xing-yang replied:
See lines 124-129, there is a `isPodTerminated`. I'm thinking `isPodTerminating` should include both `isPodTerminated` and if DeletionTimestamp is set.

jingxu97 replied:
I think in this case, we only need to check DeletionTimestamp is set and then verify whether the node is out-of-service, for other pods in terminated state, the gc controller will take care.

xing-yang replied:
Sure.

sonasingh46 replied:
I will raise a commit to address this.

sonasingh46 replied:
Done

YuikoTakada commented:
I tested this code on my environment, and it works well😄 I'm using `csi-driver-nfs` CSI driver.
I did:
1. stopped a node on which statefulset with PV running
2. added a `out-of-service` taint to the stopped node:
```
$ kubectl taint nodes test-worker2 out-of-service=value1:NoExecute
```

And then, I can see statefulset is running on the other node.

yosshy commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +128,76 @@ func isPodTerminated(pod *v1.Pod) bool {
+			taintToMatch := &v1.Taint{
+				Key:    taints.OutOfServiceTaint,
+				Effect: v1.TaintEffectNoExecute,
+			}
```
Can we move taintToMatch definition out of the loop?

yosshy commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +128,76 @@ func isPodTerminated(pod *v1.Pod) bool {
+	}
+
+	terminatedPodCount := len(terminatedPods)
+	deleteCount := terminatedPodCount
```
terminatedPodCount isn't used below, so it looks unnecessary to define both terminatedPodCount and deleteCount.

YuikoTakada commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +128,76 @@ func isPodTerminated(pod *v1.Pod) bool {
+			if !nodeutil.IsNodeReady(node) && taints.HasOutOfServiceTaintWithNoExecute(node.Spec.Taints) &&
+				!podToleratesTaint(taintToMatch, pod) {
+				terminatedPods = append(terminatedPods, pod)
+			}
```
if inclement `deleteCount` here, we don't need https://github.com/kubernetes/kubernetes/pull/108486/files#diff-eaebfd5acb747dd3dd13eb812a745b0a886ba2a24f82e852772c45ac0d936f9fR168-R169 
```suggestion
				deleteCount++
			}
```

sonasingh46 replied:
I have used the following : 
```
deleteCount := len(terminatingPods)
	if deleteCount <= 0 {
		return
	}
```
I think that should suffice.

YuikoTakada reviewed (commented):
Generally looks good to me, thanks. I left some comments, so please check them.

sonasingh46 commented:
/retest

YuikoTakada commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +129,75 @@ func isPodTerminated(pod *v1.Pod) bool {
+
+			// Add this pod to terminatedPods list only if the following conditions are met:
+			// 1. Node is not ready.
+			// 2. Node has `out-of-service` taint with `NoExecute` effect.
```
In order to avoid misunderstanding, it would be better to write the well-known taint name.
```suggestion
			// 2. Node has `node.kubernetes.io/out-of-service` taint with `NoExecute` effect.
```

YuikoTakada commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +129,75 @@ func isPodTerminated(pod *v1.Pod) bool {
+			// Add this pod to terminatedPods list only if the following conditions are met:
+			// 1. Node is not ready.
+			// 2. Node has `out-of-service` taint with `NoExecute` effect.
+			// 3. Pod does not tolerate the `out-of-service` taint with `NoExecute` effect.
```
ditto
```suggestion
			// 3. Pod does not tolerate the `node.kubernetes.io/out-of-service` taint with `NoExecute` effect.
```

YuikoTakada commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -183,8 +190,18 @@ func (rc *reconciler) reconcile() {
+
 			// Check whether volume is still mounted. Skip detach if it is still mounted unless timeout
-			if attachedVolume.MountedByNode && !timeout {
+			// or the node has `out-of-service` taint.
```
```suggestion
			// or the node has `node.kubernetes.io/out-of-service` taint.
```

YuikoTakada commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +129,75 @@ func isPodTerminated(pod *v1.Pod) bool {
+		return
+	}
+
+	klog.Infof("Garbage collecting %v pods that are terminating on node tainted with out-of-service", deleteCount)
```
ditto
```suggestion
	klog.Infof("Garbage collecting %v pods that are terminating on node tainted with node.kubernetes.io/out-of-service", deleteCount)
```

YuikoTakada reviewed (commented):
Thank you for updating. I tested latest codes on my environment and works well ;)
I did `kubectl taint nodes test-worker2 node.kubernetes.io/out-of-service=value1:NoExecute`

I wrote some comments so please check them.
BTW, this PR needs doc https://kubernetes.io/docs/reference/labels-annotations-taints/ update also? (perhaps anyone said already :p)

YuikoTakada commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +129,75 @@ func isPodTerminated(pod *v1.Pod) bool {
+	}
+
+	deleteCount := len(terminatingPods)
+	if deleteCount <= 0 {
```
`deleteCount` cannot be negative number different from the case of `gcTerminated()`
```suggestion
	if deleteCount = 0 {
```

yosshy replied:
==

YuikoTakada commented:
> #### What this PR does / why we need it:
> This PR adds a feature to detach volume immediately and does not wait for the 6 min timeout in case of a non graceful shutdown of a node that has an `out-of-service` taint added manually.

`node.kubernetes.io/out-of-service` ?

yosshy commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +129,75 @@ func isPodTerminated(pod *v1.Pod) bool {
+				Key:    v1.TaintNodeOutOfService,
+				Effect: v1.TaintEffectNoExecute,
+			}
+			if !nodeutil.IsNodeReady(node) && taints.HasOutOfServiceTaintWithNoExecute(node.Spec.Taints) &&
```
Using taints.TaintExists() looks simpler than taints.HasOutOfServiceTaintWithNoExecute() if taintToMatch is also defined here. It's an option not to use taints.HasOutOfServiceTaintWithNoExecute().
```suggestion
			if !nodeutil.IsNodeReady(node) && taints.TaintExists(node.Spec.Taints, taintToMatch) &&
```

yosshy commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -367,3 +384,13 @@ func (rc *reconciler) reportMultiAttachError(volumeToAttach cache.VolumeToAttach
+	if err != nil {
+		return false, err
+	}
+	return taints.HasOutOfServiceTaintWithNoExecute(node.Spec.Taints), nil
```
```suggestion
	taintToMatch := &v1.Taint{
		Key:    v1.TaintNodeOutOfService,
		Effect: v1.TaintEffectNoExecute,
	}
	return taints.TaintExists(node.Spec.Taints, taintToMatch), nil
```

yosshy commented on pkg/util/taints/taints.go:
```diff
@@ -275,6 +275,15 @@ func TaintExists(taints []v1.Taint, taintToFind *v1.Taint) bool {
 	return false
 }
 
+// HasOutOfServiceTaintWithNoExecute returns true if TaintNodeOutOfService with effect `NoExecute` exists
+func HasOutOfServiceTaintWithNoExecute(taints []v1.Taint) bool {
+	taintToMatch := &v1.Taint{
+		Key:    v1.TaintNodeOutOfService,
+		Effect: v1.TaintEffectNoExecute,
+	}
+	return TaintExists(taints, taintToMatch)
+}
+
```
```suggestion
```

yosshy reviewed (commented):
@sonasingh46 thank you for your PR. It works well. IMO, it's an option that you don't use taints.HasOutOfServiceTaintWithNoExecute() for readability.

sonasingh46 commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -228,8 +228,9 @@ func (rc *reconciler) reconcile() {
+			// If the node has node.kubernetes.io/out-of-service taint with NoExecute effect, skip verifySafeToDetach check
 			klog.V(5).InfoS("Starting attacherDetacher.DetachVolume", "volume", attachedVolume)
-			verifySafeToDetach := !timeout
+			verifySafeToDetach := !timeout && !hasOutOfServiceTaint
```
This change is necessary to make sure that when the pod gets on to a healthy node it starts running immediately and does not stay stuck in `ContainerCreating` state for 6 min.

thockin replied:
From a readability POV, I would suggest something like:

```
knownSafeToDetach := timeout || hasOutOfServiceTaint
err = rc.attacherDetacher.DetachVolume(attachedVolume.AttachedVolume, !knownSafeToDetach, rc.actualStateOfWorld)
```

or 

```
verifySafeToDetach := !(timeout || hasOutOfServiceTaint)
```

sonasingh46 commented:
/retest

yosshy commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -222,7 +296,6 @@ func (gcc *PodGCController) checkIfNodeExists(ctx context.Context, name string)
 // gcUnscheduledTerminating deletes pods that are terminating and haven't been scheduled to a particular node.
 func (gcc *PodGCController) gcUnscheduledTerminating(pods []*v1.Pod) {
 	klog.V(4).Infof("GC'ing unscheduled pods which are terminating.")
-
```
why removed it?

yosshy commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -324,7 +325,6 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
 	defer runtime.HandleCrash()
 	defer adc.pvcQueue.ShutDown()
 
-	klog.Infof("Starting attach detach controller")
```
why removed it?

yosshy reviewed (commented):
Thank you for your new PR. Please check my comments.

xing-yang commented:
@sonasingh46 Can you add unit tests for the changes made in gc_controller.go and attach/detach reconciler.go?

xing-yang commented:
/retest

xing-yang commented:
/retest

xing-yang commented:
/assign @mrunalp

yosshy commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -367,3 +385,17 @@ func (rc *reconciler) reportMultiAttachError(volumeToAttach cache.VolumeToAttach
+
+// hasOutOfServiceTaint returns true if the node has out of service taint present
+// with `NoExecute` effect.
+func (rc *reconciler) hasOutOfServiceTaint(nodeName types.NodeName) (bool, error) {
```
It looks cool to move it to the previous position of reconcile().

sonasingh46 replied:
I am trying to understand the reasoning behind these placements. Do we have any such standards on placing the functions? 
By the way, I have not any strong opinions here and am fine moving it. Just that I added it to below to keep the diff look simpler.

yosshy replied:
Current code has small utility functions before reconcile(), so it looks natural to insert new one like them, IMO. You can ignore my comment.

sonasingh46 replied:
Sure. I will address this. Thank you

yosshy reviewed (commented):
@sonasingh46 LGTM. I wrote a comment. Please check it.

sonasingh46 commented:
/retest

jingxu97 commented:
overall the change looks good to me. Thank you for working on it!

sonasingh46 commented:
@xing-yang @YuikoTakada @jingxu97 @yosshy @sftim -- 

I have incorporated the review comments and added test. Feel free to take a look and let me know if anything else needs to be done on this PR.
If it looks good to all of you, I can squash the commits for a final review.

jingxu97 commented:
/lgtm

xing-yang commented:
@sonasingh46 Can you resolve the merge conflict?

sonasingh46 commented:
/retest

sonasingh46 commented:
/retest

sonasingh46 commented:
/retest

jsafrane commented:
/label api-review
There is a new taint defined in this PR that needs to get API review approval.
cc @liggitt

jsafrane commented:
/sig storage
/triage accepted
/priority important-longterm

xing-yang commented:
/assign @msau42

xing-yang commented:
/milestone v1.24

endocrimes commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +130,73 @@ func isPodTerminated(pod *v1.Pod) bool {
+			defer wait.Done()
+			if err := gcc.deletePod(namespace, name); err != nil {
+				// ignore not founds
+				defer utilruntime.HandleError(err)
```
Why are you deferring the error handling here? - It'll happen immediately regardless

endocrimes commented on pkg/controller/podgc/gc_controller_test.go:
```diff
@@ -448,3 +451,149 @@ func TestGCUnscheduledTerminating(t *testing.T) {
+		},
+
+		{
+			name: "some pods has deletion timestamp and/or phase set and some of the corresponding nodes have " +
```
```suggestion
			name: "some pods have deletion timestamp and/or phase set and some of the corresponding nodes have an" +
```

endocrimes replied:
I'd also maybe just have this on one line - or break it into multiple cases. This case is a little tricky to skim.

sonasingh46 replied:
Agreed. I thought to break this one and then finally went via this route. 
My reasoning: 
I wanted to create a env where I have couple of pods distributed over couple of nodes. Plus, the nodes can be either `Ready` on `NotReady` and pods could be in diff phase. 

( Just to note that -- pod phase comes into picture -- due to the fact that we want to check if the pod is terminating or not. So we are saying the pod is terminating only when deletion timestamp is set and do not care what the phase is. There were some thoughts earlier if we want to check for pod phase or not. Currently we do not care the pod phase. )

As I was testing the top level function, `gcTerminating` -- I thought it would make sense to make it act on such env as a whole and see the output. 
( I mean, I could have broken this test case into multiple, and still inject an error in the code that would finally not be caught via the test cases. )  

Anyways, I am open to suggestions if I can make this more readable and easy to go through.

endocrimes commented on pkg/controller/podgc/gc_controller_test.go:
```diff
@@ -448,3 +451,149 @@ func TestGCUnscheduledTerminating(t *testing.T) {
+	}
+	for i, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+
```
```suggestion
```

thockin commented on pkg/features/kube_features.go:
```diff
@@ -834,6 +834,12 @@ const (
+	// alpha: v1.24
+	//
+	// Allow pods to failover to a different node in case of non graceful node shutdown
+	NonGracefulFailover featuregate.Feature = "NonGracefulFailover"
```
Can we pick a more obvious name?  NodeOutOfServiceVolumeDetach or something?

thockin commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +130,73 @@ func isPodTerminated(pod *v1.Pod) bool {
+}
+
+// podToleratesTaint returns true if the pod tolerates the provided taint.
+func podToleratesTaint(taint *v1.Taint, pod *v1.Pod) bool {
```
super nit: I'd flip these args

thockin commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +130,73 @@ func isPodTerminated(pod *v1.Pod) bool {
+}
+
+func (gcc *PodGCController) gcTerminating(pods []*v1.Pod) {
+	klog.V(4).Info("GC'ing terminating pods that are on out of service nodes")
```
super nit, I'd say "out-of-service" to be clearer that it's a specific term.

thockin commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +130,73 @@ func isPodTerminated(pod *v1.Pod) bool {
+
+			// Add this pod to terminatingPods list only if the following conditions are met:
+			// 1. Node is not ready.
+			// 2. Node has `node.kubernetes.io/out-of-service` taint with `NoExecute` effect.
```
I thought we had decided the flavor of taint did not matter, but I guess that never got resolved.  https://github.com/kubernetes/enhancements/pull/1116#discussion_r792904049

I think you don't care what taint effect is - you should only care that the pod is terminating and the node is out-of-service.

I don't think (??) you need to check whether the pod tolerates out-of-service or not - you've already confirmed it is terminating.  Is there a reason to do this extra check?

jingxu97 replied:
agree no need to check taint here. If pod has tolerations, the pod will not be evicted in the first place and is not in terminating

sonasingh46 replied:
>I don't think (??) you need to check whether the pod tolerates out-of-service or not - you've already confirmed it is terminating. Is there a reason to do this extra check?

No reason. I think I made some wrong assumptions. Thank you. 

Have incorporated the fix.

thockin commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -134,6 +141,20 @@ func (rc *reconciler) syncStates() {
+	}
+	taintToMatch := &v1.Taint{
+		Key:    v1.TaintNodeOutOfService,
+		Effect: v1.TaintEffectNoExecute,
```
again, I don't think you care the effect

sonasingh46 replied:
Done. Fixed it.

jyotimahapatra commented:
@sonasingh46 👋 I'm from 1.24 bug triage team. I wanted to remind that the [code freeze](https://github.com/kubernetes/sig-release/tree/master/release-team/role-handbooks/bug-triage#code-freeze) is approaching on 03/30.

thockin commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +130,57 @@ func isPodTerminated(pod *v1.Pod) bool {
+// isPodTerminating returns true if the pod is terminating.
+func isPodTerminating(pod *v1.Pod) bool {
+	if pod.ObjectMeta.DeletionTimestamp != nil {
+		klog.Infof("garbage collecting pod %s that is terminating. Phase [%v]", pod.Name, pod.Status.Phase)
```
Conventionally, an `isFoo` function is pretty trivial and "utility" - logging from within here seems surprising.  I think you should log this at the caller.

thockin replied:
Also - check the logs - there's ANOTHER log info below - do you need both?

thockin reviewed (commented):
All of the things I really needed fixed are fixed.  I flagged a few smaller issues.  Please check the log output when this fires and make sure it's actually useful and sufficient and not overly verbose by default.  Consider whehter more V(3) or V(4) logs would be helpful.

/approve

thockin reviewed (commented):
/approve

sonasingh46 commented:
/retest

sonasingh46 commented:
@xing-yang @mrunalp -- I squashed the commit. The verify job is failing for files that is not changed in this PR. Any ideas?

xing-yang commented:
I see similar failures on other PRs like this one: https://github.com/kubernetes/kubernetes/pull/108445.  It is not related to changes in this PR but it is blocking things from being merged.

xing-yang commented:
/retest

mrunalp commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +130,56 @@ func isPodTerminated(pod *v1.Pod) bool {
+		if isPodTerminating(pod) {
+			node, err := gcc.nodeLister.Get(string(pod.Spec.NodeName))
+			if err != nil {
+				klog.Errorf("failed to get node:%s", err)
```
nit: node: %s

mrunalp commented:
This looks fine to me. 
 @gnufied want to take a look as well before we merge?

jingxu97 commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -211,8 +238,9 @@ func (rc *reconciler) reconcile() {
+			verifySafeToDetach := !(timeout || hasOutOfServiceTaint)
 			err = rc.attacherDetacher.DetachVolume(attachedVolume.AttachedVolume, verifySafeToDetach, rc.actualStateOfWorld)
 			if err == nil {
 				if !timeout {
```
add log for the case of hasOutOfServiceTaint here.

sonasingh46 replied:
done

gnufied commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -124,6 +130,56 @@ func isPodTerminated(pod *v1.Pod) bool {
+	if pod.ObjectMeta.DeletionTimestamp != nil {
+		return true
+	}
+	return false
```
nit: you can also do `return pod.ObjectMeta.DeletionTimestamp != nil`

gnufied commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -183,8 +200,18 @@ func (rc *reconciler) reconcile() {
 			timeout := elapsedTime > rc.maxWaitForUnmountDuration
+
+			hasOutOfServiceTaint := false
+			if utilfeature.DefaultFeatureGate.Enabled(features.NodeOutOfServiceVolumeDetach) {
```
nit: can you move feature gate check inside `hasOutOfServiceTaint` function, so as it always returns false if feature gate is not enabled?

sonasingh46 commented:
@mrunalp @gnufied @jingxu97 -- Have addressed the comments and raised a commit. I will squash it once you review and it looks good to you all.

jingxu97 commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -211,8 +238,12 @@ func (rc *reconciler) reconcile() {
 			klog.V(5).InfoS("Starting attacherDetacher.DetachVolume", "volume", attachedVolume)
-			verifySafeToDetach := !timeout
+			if hasOutOfServiceTaint {
+				klog.V(5).Infof("node %q has out-of-service taint", attachedVolume.NodeName)
```
I think we can use level 4 log here.

sonasingh46 replied:
Sure. Changed it. 
Also, trying to learn here: How do we decide on when to use what log level when we have around 1-10? -- Like I tried to find some pattern in the existing code base but could not grok much.

jingxu97 replied:
Just my typical use of logging level. For very useful and critical log, set it to level 2 or below. For useful log for debugging purpose, we can set it level 4. I rarely use 5 or above since we typically do not set logging to that high.

sonasingh46 replied:
Thank you. 🙏  This helps.

mrunalp commented:
:+1:

sonasingh46 commented:
I have tested the changes and the tests works as expected which is described in the PR description.

mrunalp commented:
/lgtm

xing-yang commented:
/retest

xing-yang commented:
@sonasingh46 There's a merge conflict.  Can you take a look?

xing-yang commented:
/retest

sonasingh46 commented:
/retest

xing-yang commented:
/lgtm

xing-yang commented:
/retest

sonasingh46 commented:
/retest

tengqm commented:
The new feature gate needs a docs website update.

xing-yang commented:
Thanks @tengqm for the reminder.  
@sonasingh46 Can you add the feature gate in your doc PR?
https://github.com/kubernetes/website/pull/32406
